src/server/bitops_family.cc

```diff
@@ -577,12 +593,488 @@ void BitCount(CmdArgList args, ConnectionContext* cntx) {
   HandleOpValueResult(res, cntx);
 }
 
+enum class EncodingType { UINT, INT, NILL };
+
+struct CommonAttributes {
+  EncodingType type;
+  size_t encoding_bit_size;
+  size_t offset;
+};
+
+struct Nill {};
+using ResultType = std::variant<int64_t, Nill>;
+
+struct Overflow {
+  enum Policy { WRAP, SAT, FAIL };
+
+  static std::optional<Policy> PolicyFromStr(std::string_view policy) {
+    if (policy == "WRAP") {
+      return WRAP;
+    }
+    if (policy == "SAT") {
+      return SAT;
+    }
+    if (policy == "FAIL") {
+      return FAIL;
+    }
+    return {};
+  }
+
+  bool UIntOverflow(int64_t* value, int64_t incr, size_t total_bits) const {
+    // total up to 63 bits -- we do not support 64 bit unsigned
+    const uint64_t max = (1UL << total_bits) - 1;
+
+    uint64_t incr_value = incr;
+    if (incr_value + *value > max) {
+      switch (type) {
+        case Overflow::WRAP:
+          // safe to do, won't overflow, both incr and value are <= than 2^63 - 1
+          *value = (incr_value + *value) % max;
+          break;
+        case Overflow::SAT:
+          *value = max;
+          break;
+        case Overflow::FAIL:
+          *value = 0;
+          return false;
+      }
+      return true;
+    }
+
+    *value = incr_value + *value;
+    return true;
+  }
+
+  bool IntOverflow(int64_t* value, size_t total_bits, int64_t incr = 0, bool add = false) const {
```

Comment:
I personally prefer that long functions will be defined outside of the class definition to improve readability.
Also, output parameter.

Reply:
done